src/bucket/BucketIndex.cpp

```diff
@@ -0,0 +1,339 @@
+static LedgerKey
+getDummyPoolShareTrustlineKey(AccountID const& accountID, uint8_t fill)
+{
+    LedgerKey key(TRUSTLINE);
+    key.trustLine().accountID = accountID;
+    key.trustLine().asset.type(ASSET_TYPE_POOL_SHARE);
+    key.trustLine().asset.liquidityPoolID().fill(fill);
+    return key;
+}
+
+// Index maps a range of BucketEntry's to the associated offset
+// within the bucket file. Index stored as vector of pairs:
+// First: LedgerKey/Key ranges sorted in the same scheme as LedgerEntryCmp
+// Second: offset into the bucket file for a given key/ key range.
+// pageSize determines how large, in bytes, each range should be. pageSize == 0
+// indicates individual keys used instead of ranges.
+template <class IndexT> class BucketIndexImpl : public BucketIndex
+{
+    IndexT mKeysToOffset{};
+    std::streamoff const mPageSize{};
+    std::unique_ptr<bloom_filter> mFilter{};
+
+    BucketIndexImpl(BucketManager const& bm,
+                    std::filesystem::path const& filename,
+                    std::streamoff pageSize);
+
+    friend std::unique_ptr<BucketIndex const>
+    BucketIndex::createIndex(BucketManager const& bm,
+                             std::filesystem::path const& filename);
+
+  public:
+    BucketIndexImpl() = default;
+
+    virtual std::optional<std::streamoff>
+    lookup(LedgerKey const& k) const override;
+
+    virtual std::pair<std::optional<std::streamoff>, Iterator>
+    scan(Iterator start, LedgerKey const& k) const override;
+
+    virtual std::pair<std::streamoff, std::streamoff>
+    getPoolshareTrustlineRange(AccountID const& accountID) const override;
+
+    virtual std::streamoff
+    getPageSize() const override
+    {
+        return mPageSize;
+    }
+
+    virtual Iterator
+    begin() const override
+    {
+        return mKeysToOffset.begin();
+    }
+
+    virtual Iterator
+    end() const override
+    {
+        return mKeysToOffset.end();
+    }
+};
+
+template <class IndexT>
+BucketIndexImpl<IndexT>::BucketIndexImpl(BucketManager const& bm,
+                                         std::filesystem::path const& filename,
+                                         std::streamoff pageSize)
+    : mPageSize(pageSize)
+{
+    ZoneScoped;
+    releaseAssert(!filename.empty());
+
+    auto timer = LogSlowExecution("Indexing bucket");
+
+    size_t const estimatedLedgerEntrySize =
+        xdr::xdr_traits<BucketEntry>::serial_size(BucketEntry{});
+    auto fileSize = fs::size(filename.string());
+    auto estimatedNumElems = fileSize / estimatedLedgerEntrySize;
+    size_t estimatedIndexEntries;
+
+    // Initialize bloom filter for range index
+    if constexpr (std::is_same<IndexT, RangeIndex>::value)
+    {
+        ZoneNamedN(bloomInit, "bloomInit", true);
+        bloom_parameters params;
+        params.projected_element_count = estimatedNumElems;
+        params.false_positive_probability = 0.001; // 1 in 1000
+        crypto_shorthash_keygen(params.random_seed.data());
```

Comment:
Unfortunately this is setting the random seed using real entropy which makes the program nondeterministic in testing. Need to use `gRandomEngine`.

Reply:
Or maybe use the same `gKey` that we use in shorthash.cpp ?